url6a9f553762_Bresiu_socket_io_tgz-pJ8-MainJ8/src/com/github/nkzawa/utf8/UTF8.java

```diff
@@ -0,0 +1,164 @@
+package com.github.nkzawa.utf8;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * UTF-8 encoder/decoder ported from utf8.js.
+ *
+ * @see <a href="https://github.com/mathiasbynens/utf8.js">https://github.com/mathiasbynens/utf8.js</a>
+ */
+public class UTF8 {
+
+    private static int[] byteArray;
+    private static int byteCount;
+    private static int byteIndex;
+
+    public static String encode(String string) {
+        int[] codePoints = uc2decode(string);
+        int length = codePoints.length;
+        int index = -1;
+        int codePoint;
+        StringBuilder byteString = new StringBuilder();
+        while (++index < length) {
+            codePoint = codePoints[index];
+            byteString.append(encodeCodePoint(codePoint));
+        }
+        return byteString.toString();
+    }
+
+    public static String decode(String byteString) throws UTF8Exception {
+        byteArray = uc2decode(byteString);
+        byteCount = byteArray.length;
+        byteIndex = 0;
+        List<Integer> codePoints = new ArrayList<Integer>();
+        int tmp;
+        while ((tmp = decodeSymbol()) != -1) {
+            codePoints.add(tmp);
+        }
+        return ucs2encode(listToArray(codePoints));
+    }
+
+    private static int[] uc2decode(String string) {
+        int length = string.length();
+        int[] output = new int[string.codePointCount(0, length)];
+        int counter = 0;
+        int value;
+        for (int i = 0; i < length; i += Character.charCount(value)) {
+            value = string.codePointAt(i);
+            output[counter++] = value;
+        }
+        return output;
+    }
+
+    private static String encodeCodePoint(int codePoint) {
+        StringBuilder symbol = new StringBuilder();
+        if ((codePoint & 0xFFFFFF80) == 0) {
+            return symbol.append(Character.toChars(codePoint)).toString();
+        }
+        if ((codePoint & 0xFFFFF800) == 0) {
+            symbol.append(Character.toChars(((codePoint >> 6) & 0x1F) | 0xC0));
+        } else if ((codePoint & 0xFFFF0000) == 0) {
+            symbol.append(Character.toChars(((codePoint >> 12) & 0x0F) | 0xE0));
+            symbol.append(createByte(codePoint, 6));
+        } else if ((codePoint & 0xFFE00000) == 0) {
+            symbol.append(Character.toChars(((codePoint >> 18) & 0x07) | 0xF0));
+            symbol.append(createByte(codePoint, 12));
+            symbol.append(createByte(codePoint, 6));
+        }
+        symbol.append(Character.toChars((codePoint & 0x3F) | 0x80));
+        return symbol.toString();
+    }
+
+    private static char[] createByte(int codePoint, int shift) {
+        return Character.toChars(((codePoint >> shift) & 0x3F) | 0x80);
+    }
+
+    private static int decodeSymbol() throws UTF8Exception {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 99-107, lines 109-118, lines 120-128 into separate methods.